src/runtime/task/async_task.rs

```diff
@@ -0,0 +1,43 @@
+use crate::runtime::execution::ExecutionState;
+use crate::runtime::thread::continuation::{ContinuationPool, PooledContinuation};
+use crate::runtime::{task::*, thread};
+use futures::future::BoxFuture;
+use std::task::{Context, Poll};
+
+pub(crate) struct AsyncTask {
+    continuation: PooledContinuation,
+}
+
+pub(crate) fn new(thread_stack_size: usize, mut future: BoxFuture<'static, ()>) -> AsyncTask {
+    let mut continuation = ContinuationPool::acquire(thread_stack_size);
+    continuation.initialize(Box::new(move || {
+        // TODO should I be updating the waker on each iteration?
+        let waker = ExecutionState::with(|state| state.current_mut().waker());
+        let cx = &mut Context::from_waker(&waker);
+        while future.as_mut().poll(cx) == Poll::Pending {
+            ExecutionState::with(|state| {
+                // Want block_after_polling here because block_after_running will later be called
+                // Need to block here before thread::switch (if we didn't wake ourselves)
+                // BUT we don't want the execution loop to block with block_after_running in the case that we did wake ourselves
+                state.current_mut().block_after_polling();
```

Comment:
(Commenting summary of our discussion for completeness)
We need to block before `thread::switch` because the scheduler may choose to continue the same task and avoid a context switch as an optimization.
Since `block_after_running` is called during this context switch, it will be skipped due to the optimization.
Previously, this wasn't an issue because futures could not call `thread::switch`.
The overall solution seems to be removing `block_after_running` from the context switch and instead requiring that futures mark themselves as blocked before calling `thread::switch`.